### PR TITLE
Lazily pack and unpack textures in WebGL kernels

### PR DESCRIPTION
**Pull request by annxingyuan**
### Changes
- Add `packedInput` boolean property to GPGPUProgram to identify WebGL kernels that require packed inputs (`batchMatMul`, `conv2DWithIm2Col`).
- Such kernels no longer unpack their outputs, and they only pack their inputs if they are not already packed.
- In `compileAndRun`, check whether the GPGPUProgram requires packed inputs. If it does not, and an input texture is packed, unpack it before passing it to the program.

PERF

---

##### For repository owners only:

Please remember to apply all applicable tags to your pull request.
Tags: FEATURE, BREAKING, BUG, PERF, DEV, DOC, SECURITY

For more info see: https://github.com/tensorflow/tfjs/blob/master/DEVELOPMENT.md

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1329)

### CODE REVIEW

**Comment by annxingyuan** [comment]
@nsthorat 

If I dispose the input tensors to a kernel immediately after packing / unpacking them, gradient operations stop working because the inputs to `engine.runKernel` are disposed by the time they get passed to the tape node. 

Maybe it would make sense to have the gradient scope dispose of the tensors instead? But I'm sure how to approach implementing that - is there a straightforward way that takes advantage of existing infrastructure that you know of?

**Comment by annxingyuan** [comment]
Note - this PR now depends on https://github.com/tensorflow/tfjs-core/pull/1327 because the batchNorm tests need to be able to print 4D packed outputs
